src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,614 @@
+  *
+  * A [[SAMFileHeader]] should be given per sample and a [[ReadStructure]] per read from the same template/fragment.
+  * Use the [[demultiplex()]] method to create a [[SAMRecord]] for each read with template bases.  Any molecular barcodes
+  * will be extracted and stored in the tag specified by [[umiTag]].
+  *
+  * @param headers a sequence of [[SAMFileHeader]]s for each sample in the same order as samples in the sample sheet.
+  * @param samples the samples ordered by sample ordinal.
+  * @param readStructures the read structures, one for each read that will be given to [[demultiplex()]].
+  * @param umiTag the tag to store any molecular barcodes.  The barcodes from reads will be delimited by "-".
+  * @param qualityFormat the quality format for the FASTQ records.
+  * @param minQ the minimum base quality to enforce.
+  * @param maxQ the maximum base quality to enforce.
+  * @param maxMismatches the maximum mismatches to match a sample barcode.
+  * @param minMismatchDelta the minimum difference between number of mismatches in the best and second best barcodes for
+  *                         a barcode to be considered a match.
+  * @param maxNoCalls the maximum number of no calls in the sample barcode bases allowed for matching.
+  */
+private class FastqDemultiplexer(val headers: Seq[SAMFileHeader],
+                                 val samples: Seq[Sample],
+                                 val readStructures: Seq[ReadStructure],
+                                 val umiTag: String = "RX",
+                                 val qualityFormat: FastqQualityFormat = FastqQualityFormat.Standard,
+                                 val minQ: Int = 0,
+                                 val maxQ: Int =  SAMUtils.MAX_PHRED_SCORE,
+                                 val maxMismatches: Int = 2,
+                                 val minMismatchDelta: Int = 1,
+                                 val maxNoCalls: Int = 2) {
+  import FastqDemultiplexer._
+
+  require(readStructures.nonEmpty, "No read structures were given")
+  require(samples.map(_.sampleOrdinal).toList == samples.map(_.sampleOrdinal).sorted, "Samples not provided in ascending order by sample ordinal")
+  require(samples.flatMap(_.sampleBarcode).map(_.concatenatedBarcode).sorted.distinct.length == samples.length, "Unique sample barcodes required for all samples")
+  require(headers.length == samples.length + 1, s"Not enough SAMFileHeaders; found ${headers.length} but should have ${samples.length+1}.")
+  require(headers.forall(_.getReadGroups.size == 1), "All headers must have one read group. Ordinals are: " + headers.zipWithIndex.filter(_._1.getReadGroups.size != 1).map(_._2).mkString(", "))
+
+  /** The number of reads that are expected to be given to the [[demultiplex()]]. */
+  def expectedNumberOfReads: Int = readStructures.length
+
+  /** True if we the read structure implies paired end reads will be produced, false otherwise. */
```

Comment:
change to: `True if the read structure implies paired end reads will be produced, false otherwise`